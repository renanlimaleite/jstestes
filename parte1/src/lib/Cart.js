export const Cart = () => {
  let items = []

  const getTotal = () => {
    return items.reduce((acc, item) => acc + item.product.price * item.quantity, 0)
  }

  const add = (item) => {
    const find = items.find(p => p.product.title === item.product.title)
    if (!find) {   
      items = [...items, item]
    }
  }

  const remove = (productArg) => {
    items = items.filter(prd => prd.product.title !== productArg.title)
  }

  const summary = () => {}

  const checkout = () => {
    return {
      total: getTotal(),
      items
    }
  }

  return {
    items,
    getTotal,
    add,
    remove,
    summary,
    checkout
  }
}

const cart = Cart()

const product = {
  title: '',
  price: ''
}

const item = {
  quantity: 2,
  product
}

cart.add(item)
cart.remove(product)
cart.getTotal()
cart.summary()
cart.checkout()
