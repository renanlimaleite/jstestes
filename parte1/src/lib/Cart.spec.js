import { Cart } from './Cart'

describe('Cart', () => {
  let cart

  let product = {
    title: 'Adidas running shoes - men',
    price: 23488 // 234.88 | R$ 234,88
  }
  let product2 = {
    title: 'Adidas running shoes - women',
    price: 20000 // 234.88 | R$ 234,88
  }

  beforeEach(() => {
    cart = Cart()
  })

  describe('getTotal()', () => {
    
  it('should return 0 when getTotal() is executed in a newly created instance', () => {
    expect(cart.getTotal()).toEqual(0)
  });

  it('should multiply quantity and price receive the total amount', () => {
    const item = {
      product,
      quantity: 2, // 46976
    }

    cart.add(item)

    expect(cart.getTotal()).toEqual(46976)
  });
  
  it('should ensure no more than on product exists at a time', () => {
    cart.add({
      product,
      quantity: 2
    })

    cart.add({
      product,
      quantity: 1
    })

    expect(cart.getTotal()).toEqual(46976)
  });

  it('should update total when a products get included and then removed', () => {
    cart.add({
      product,
      quantity: 2
    })

    cart.add({
      product: product2,
      quantity: 1
    })

    cart.remove(product)

    expect(cart.getTotal()).toEqual(20000)
  });
  });
  
  describe('checkout()', () => {
    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 5
      })

      cart.add({
        product: product2,
        quantity: 3
      })

      // toMatchInlineSnapshot
      expect(cart.checkout()).toMatchSnapshot()
    });
  });
});