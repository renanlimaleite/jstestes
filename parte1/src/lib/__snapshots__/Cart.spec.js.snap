// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 23488,
        "title": "Adidas running shoes - men",
      },
      "quantity": 5,
    },
    Object {
      "product": Object {
        "price": 20000,
        "title": "Adidas running shoes - women",
      },
      "quantity": 3,
    },
  ],
  "total": 177440,
}
`;
